File: javascript-sdk/packages/core/src/payment-client.ts
import { Repository } from "./services/repository";
import { SocketService } from "./services/socket_service";
import { Asset } from "./types/asset";
import { PaymentOptions, QrRequestDetails, QrResponseData } from "./types/payment-client-types";

// --- Clase PaymentClient ---
export class PaymentClient {
  private options: PaymentOptions;
  private socket: SocketService;

  constructor(options: PaymentOptions) {
    this.options = options;
    this.socket = new SocketService(this.options);
  }

  // --- Métodos para obtener datos ---
  public async getAssets(): Promise<Asset[]> {
    const assets = await Repository.getAssets();

    return assets.result!;
  }

  // --- Método para obtener datos del QR ---
  public async fetchQrCodeDetails(details: QrRequestDetails): Promise<QrResponseData> {
    const qrData = await Repository.fetchQrCodeDetails({
      ...details,
      amount: this.options.amount,
      metadata: this.options.metadata,
      publicKey: this.options.publicKey
    });

    console.log('qrData', qrData);

    this.socket.connect(qrData.result!.id);

    return qrData.result!;
  }

  public disconnectWebSocket(): void {
    this.socket.disconnect();
  }
}

File: dart_sdk/lib/services/payment_client.dart
import 'package:payment_button_sdk/models/asset.dart';
import 'package:payment_button_sdk/models/payment_client_models.dart';

import 'repository.dart';
import 'socket_service.dart';

class PaymentClient {
  final PaymentOptions options;
  late final SocketService _socketService;

  PaymentClient(this.options) {
    _socketService = SocketService(options);
  }

  Future<List<Asset>> getAssets() async {
    final response = await Repository.getAssets();
    return response.result!;
  }

  Future<QrResponseData> fetchQrCodeDetails({
    required String assetId,
    required String networkId,
  }) async {
    final qrData = await Repository.fetchQrCodeDetails(
      publicKey: options.publicKey,
      amount: options.amount,
      assetId: assetId,
      networkId: networkId,
      metadata: options.metadata,
    );

    _socketService.connect(qrData.result!.id);

    return qrData.result!;
  }

  void disconnectWebSocket() {
    _socketService.disconnect();
  }
}
